doc/porgrc.5.in
.\"----------------------------------------------------------------
.\" porgrc.8 - porgrc man page
.\"----------------------------------------------------------------
.\" Copyright (C) 2014 David Ricart <@PACKAGE_BUGREPORT@>
.\"----------------------------------------------------------------

.TH porgrc "5" "@RELEASEDATE@" "porg @PACKAGE_VERSION@" "System"
.SH NAME
porgrc - configuration file for porg
.SH SYNOPSIS
\fB@sysconfdir@/porgrc\fR
.SH DESCRIPTION
.PP
\fBporgrc\fR is the system-wide configuration file for \fBporg(8)\fR,
the source code package organizer.
.SH SYNTAX
.PP
The file consists of comments and key-value pairs. Comments are on separate
lines that start with a \fB#\fR symbol. The key-value pairs are on the form:
.PP
\fBkey=value\fR
.PP
where \fBkey\fR contains no whitespace and no equal signs, and
\fBvalue\fR will be all of the text after the equal sign until (but not
including) the ending newline.
.SH OPTIONS
The following options can be set. The equivalent command line options (if any),
are listed in brackets:
.TP
\fBLOGDIR\fR [-L|--logdir]
.br
Base porg logs directory. Default is '@LOGDIR@'. Environment variables like
$HOME are expanded at runtime.
.TP
\fBINCLUDE\fR [-I|--include]
.br
Colon-separated list of paths to scan when logging the installation of a
package. Default is '/'.
.br
Shell wildcards are allowed in the paths. See \fIPATH MATCHING\fR below for
more details.
.TP
\fBEXCLUDE\fR [-E|--exclude]
.br
Colon-separated list of paths to skip when logging the installation of a
package. Default is '@EXCLUDE@'.
.br
Shell wildcards are allowed in the paths. See \fIPATH MATCHING\fR below for
more details.
.TP
\fBREMOVE_SKIP\fR [-e|--skip]
.br
Colon-separated list of paths to skip when removing a
package. Default is '' (all logged files are removed).
.br
Shell wildcards are allowed in the paths. See \fIPATH MATCHING\fR below for
more details.
.SH PATH MATCHING
Variables \fB\INCLUDE\fR, \fBEXCLUDE\fR and \fBREMOVE_SKIP\fR accept a 
colon-separated list of
paths, each of which may contain shell-like wildcards (*, ? and [..]).
Files are matched against each of those paths, following the standard
shell-like expansion, but with the following exception: If a path in the list
does not contain any wildcard, and it is a directory, it matches any file
within that directory.
.SH FILES
.PP
\fI@sysconfdir@/porgrc\fR - active configuration file
.br
\fI@prefix@/share/porg/porgrc\fR - sample configuration file
.SH AUTHOR
Written by David Ricart (@PACKAGE_BUGREPORT@)
.SH SEE ALSO
porg(8), porgball(8)


ChangeLog
SVN
---

	+ Save grop settings in ~/.config/grop/groprc instead of ~/.grop
	+ Improved the code for checking for command line option compatibility.


Version 0.4 (26/05/2014)
------------------------

	+ Bugfix: In some cases, tmp files were not removed after installing a 
	  package.
	+ Bugfix: Some files were not registered when logging the own porg
	  installation with "make logme".
	+ Bugfix: The main window of grop was partially invisible.
	+ Bugfix: Porgball failed to extract porgballs in some cases.
	+ New feature: When installing a package, save the environment variables
	  passed to ./configure, and not only the command line options.
	+ New feature: Expand any environment variable (and not only $HOME),
	  in variable LOGDIR in porgrc.
	+ New feature: Added option '-X|--exact-version' to porg and porgball (see 
	  man page).
	+ Improved the code for searching for package information.
	+ Added porg.spec.
	+ Removed the (buggy) superporg script.
	+ Added short option '-m' for '--log-missing'.


Version 0.3 (19/05/2014)
------------------------

	+ Readded GPL License and Copyright.
	+ Bugfix: Properly log files with spaces in their names.
	+ Bugfix: EXCLUDE variable in porgrc was not parsed correctly.
	+ Bugfix: Skip sourcing of /etc/bash_completion when installing
	  porg_bash_completion, to avoid errors from foreign software.


Version 0.2 (23/04/2014)
------------------------

	+ Several bugs fixed thanks to the contribution of Masahiro Kasahara:
	  - paco2porg failed when importing paco logs with no logged files
	  - porg hanged up when logging packages and the installer program calls 
	    jemalloc
	+ Fixed bug that prevented logging of files created by programs that don't
	  terminate normally.
	+ Fixed bug with resolution of relative paths.
	+ Removed option '-g|--log-ignore-errors'.
	+ Handle system calls openat(), linkat(), etc... when installing a package.
	+ Do not handle *truncate() system calls when installing a package.
	+ Removed porg.pc from distribution, as porg is not a library.


Version 0.1 (16/03/2014)
------------------------

	+ Removed question 3 of the FAQ, which discusses logging of directories.
	+ Removed getopt from the distribution. Use native getopt.
	+ Do not check for regex.h in configure (assume it is available).
	+ Rewritten porgball in bash, and fixed some bugs in it.
	+ Fixed bug in porg when installing packages with description spanning
	  more than one line (thanks Gordon Findlay).


Version 0.0 (14/03/2014)
------------------------

Changes from the last version of paco:

	+ Disabled the options for removing shared files when uninstalling a
	  package, both in porg and grop. Now shared files are never removed,
	  as it ougth to be.
	+ Disabled listing of missing and shared files.
	+ Simplification of the GUI.
	+ Major code enhancements and cleanup.



lib/porg/porgrc.cc
//=======================================================================
// porgrc.cc
//-----------------------------------------------------------------------
// This file is part of the package porg
// Copyright (C) 2014 David Ricart
// For more information visit http://porg.sourceforge.net
//=======================================================================

#include "config.h"
#include "porgrc.h"
#include "rexp.h"
#include <wordexp.h>
#include <fstream>

using std::string;
using namespace Porg;

static string sh_expand(string const&);


namespace Porg
{
	string Porgrc::s_logdir = LOGDIR;
	string Porgrc::s_include = "/";
	string Porgrc::s_exclude = EXCLUDE;
	string Porgrc::s_remove_skip = string();
}


Porgrc::Porgrc()
{
	std::ifstream f(PORGRC);
	if (!f)
		return;

	Rexp re("^([A-Z_]+)=(.*)$");

	for (string buf, opt, val; getline(f, buf); ) {
		if (re.exec(buf)) {
			opt = re.match(1);
			val = re.match(2);
			if (opt == "LOGDIR")
				s_logdir = sh_expand(val);
			else if (opt == "INCLUDE")
				s_include = val;
			else if (opt == "EXCLUDE")
   				s_exclude = val;
			else if (opt == "REMOVE_SKIP")
   				s_remove_skip = val;
		}
	}
}


bool Porgrc::logdir_writable()
{
	return !access(s_logdir.c_str(), W_OK);
}


static string sh_expand(string const& str)
{
	wordexp_t p;

	if (0 != wordexp(str.c_str(), &p, WRDE_NOCMD))
		return str;
 
 	string ret = p.we_wordv[0];
	for (uint i = 1; i < p.we_wordc; ++i)
		ret += string(" ") + p.we_wordv[i];

	wordfree(&p);
  	return ret;
}

